app/controllers/case_information_controller.rb

```diff
@@ -5,12 +5,67 @@ class CaseInformationController < PrisonsApplicationController
   before_action :set_prisoner_from_url, only: [:new, :edit, :edit_prd, :update_prd, :show]
 
   def new
-    @case_info = CaseInformation.new(
+    @last_location = LastLocationForm.new(
       nomis_offender_id: nomis_offender_id_from_url
     )
   end
 
-  def edit; end
+  def create
+    if params[:stage] == 'last_location'
+      @last_location = LastLocationForm.new(last_location_params)
+      @case_info = CaseInformation.new(
+        nomis_offender_id: @last_location.nomis_offender_id,
+        probation_service: @last_location.probation_service,
+        manual_entry: true
+      )
+      if @last_location.valid?
+        unless @case_info.valid?
+          @prisoner = prisoner(@last_location.nomis_offender_id)
+          render :missing_info
+        end
+      else
+        display_address_page
+      end
+    else
+      @last_location = Struct.new(:valid?).new(true)
+      @case_info = CaseInformation.new(case_information_params.merge(team_id: team_identifier, manual_entry: true))
+      unless @case_info.valid?
+        handle_stage2_rendering
+      end
+    end
+    if @last_location.valid? && @case_info.valid?
+      @case_info.save!
+      send_email
+      redirect_to prison_summary_pending_path(active_prison_id, sort: params[:sort], page: params[:page])
+    end
```

Comment:
I think there are could be a few code smells in this method which would be nice to clear up.

A few things which stand out to me:

- Naming between the steps/stages of the form is inconsistent. We have two stages – sometimes called 'stage2' (e.g. method `handle_stage2_rendering`), and sometimes referred to by name (e.g. `params[:stage] == 'last_location'`).
  - Stage 1 is referred to as:
    - 'last location' – i.e. we receive a param `params[:stage] == 'last_location'`
    - 'address page' – i.e. `display_address_page` method
    - The view for this stage is `new.html.erb` – which isn't immediately obvious that it's for this stage
  - Stage 2 is referred to as:
    - 'missing info' – i.e. `render :missing_info`
    - 'stage 2' – i.e. `handle_stage2_rendering`
    - The view for this stage is `missing_info.html.rb` – out of context, this might read like the name of an error page. Perhaps `add_missing_info` would be a more explicit name.
- I find the control flow in this method quite hard to follow. I wonder if there's any way we could simplify this.
- It's not immediately obvious which stage of the form comes first when we refer to them by name. I'm not sure if this could become clearer through simplified control flow, or maybe just a comment in the method.
- Smell: It feels a bit hacky that the final save (line 37) relies upon always having a truthy `@last_location.valid?` – so in stage 2 (line 30) we actually have to create a fake LastLocationForm object which returns true. I wonder if a change in control flow could help here, too.

This could be worth a conversation to talk it through – or I'd also be happy to take a crack at seeing if I can make any improvements myself.

Reply:
Agreed. This is something I had considered. Maybe having a 'stage1' method in the routing, so that we hit that rather than 'create' from the first form?